src/rpc/context.rs

```diff
@@ -0,0 +1,361 @@
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        username: String,
+        password: String,
+        address: Option<Address<N>>,
+        peers: Arc<Peers<N, E>>,
+        ledger: LedgerReader<N>,
+        operator: Arc<Operator<N, E>>,
+        prover_router: ProverRouter<N>,
+        memory_pool: Arc<RwLock<MemoryPool<N>>>,
+    ) -> Self {
+        Self(Arc::new(RpcInner {
+            address,
+            peers,
+            ledger,
+            operator,
+            prover_router,
+            memory_pool,
+            credentials: RpcCredentials { username, password },
+            launched: Instant::now(),
+        }))
+    }
+}
+
+/// Defines the authentication format for accessing private endpoints on the RPC server.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct RpcCredentials {
+    /// The username in the credential
+    pub username: String,
+    /// The password in the credential
+    pub password: String,
+}
+
+/// RPC metadata for encoding authentication.
+#[derive(Default, Clone)]
+pub struct Meta {
+    /// An optional authentication string for protected RPC functions.
+    pub auth: Option<String>,
+}
+
+/// Starts a local RPC HTTP server at `rpc_port` in a dedicated `tokio` task.
+/// RPC failures do not affect the rest of the node.
+pub async fn initialize_rpc_server<N: Network, E: Environment>(
+    rpc_server_addr: SocketAddr,
+    rpc_server_context: RpcContext<N, E>,
+) -> (SocketAddr, tokio::task::JoinHandle<()>) {
+    let access_control = AccessControlBuilder::default().build(); // TODO(ljedrz): adjust to only accept the desired hosts.
+
+    let server = HttpServerBuilder::new()
+        .set_access_control(access_control)
+        .max_request_body_size(10 * 1024 * 1024) // Explicitly select the body size limit (jsonrpsee's default, 10MiB) for greater visibility.
+        .build(rpc_server_addr).expect("Failed to create the RPC server");
+
+    let server_addr = server.local_addr().expect("Can't obtain RPC server's local address");
+
+    let module = create_rpc_module(rpc_server_context).expect("Failed to start the RPC server");
+
+    let (router, handler) = oneshot::channel();
+    let task = tokio::spawn(async move {
+        // Notify the outer function that the task is ready.
+        let _ = router.send(());
+        let server_handle = server.start(module).expect("Failed to start the RPC server");
+        server_handle.await
+    });
+    // Wait until the spawned task is ready.
+    let _ = handler.await;
+
+    (server_addr, task)
+}
+
+fn create_rpc_module<N: Network, E: Environment>(rpc_context: RpcContext<N, E>) -> Result<RpcModule<RpcContext<N, E>>, JsonrpseeError> {
+    let mut module = RpcModule::new(rpc_context);
+
+    // Public methods.
+
+    module.register_async_method("latestblock", |_rpc_params, rpc_context| async move {
+        rpc_context.latest_block().map_err(|e| JsonrpseeError::Request(e.to_string())).await
```

Comment:
the intention is that you should use https://docs.rs/jsonrpsee-core/latest/jsonrpsee_core/error/enum.Error.html#method.to_call_error here if you don't care about the specific JSON-RPC error code and optional data.